apps/mark-scan/backend/src/custom-paper-handler/state_machine.ts

```diff
@@ -0,0 +1,266 @@
+   * no paper to eject, handler will do nothing and return positive acknowledgement.
+   */
+  async ejectPaper(): Promise<void> {
+    this.status = 'ejecting';
+    await this.driver.ejectPaper();
+  }
+
+  /**
+   * Moves paper to print position and moves print head to DOWN position. If
+   * paper is already in an appropriate print position, does not move paper.
+   * E.g. if paper is loaded, it will pull the paper in a few more inches, but
+   * if the paper is parked, it will not move the paper. Printing can start in
+   * a variety of positions.
+   */
+  async enablePrint(): Promise<void> {
+    await this.driver.enablePrint();
+  }
+
+  /**
+   * Moves print head to UP position, does not move paper
+   */
+  async disablePrint(): Promise<void> {
+    await this.driver.disablePrint();
+  }
+
+  async scanAndSave(pathOut: string): Promise<void> {
+    await this.driver.setScanDirection('backward');
+    const grayscaleResult = await this.driver.scan();
+    const colorResult = new Uint8Array(grayscaleResult.byteLength * 4);
+    for (let i = 0; i < grayscaleResult.byteLength; i += 1) {
+      colorResult.set(
+        [
+          grayscaleResult.at(i) as Uint8,
+          grayscaleResult.at(i) as Uint8,
+          grayscaleResult.at(i) as Uint8,
+          255,
+        ],
+        i * 4
+      );
+    }
+    const imageData = createImageData(
+      Uint8ClampedArray.from(colorResult),
+      1728,
+      grayscaleResult.byteLength / 1728
+    );
+    await writeImageData(pathOut, imageData);
+  }
+
+  async printBallot(
+    pdfData: Uint8Array,
+    options: Partial<ImageConversionOptions> = {}
+  ): Promise<void> {
+    this.status = 'printing_ballot';
+    const enablePrintPromise = this.enablePrint();
+
+    let time = Date.now();
+    const pages: ImageData[] = [];
+    for await (const { page, pageCount } of pdfToImages(Buffer.from(pdfData), {
+      scale: 200 / 72,
+    })) {
+      assert(pageCount === 1, `Unexpected page count ${pageCount}`);
+      pages.push(page);
+    }
+    const page = pages[0];
+    assert(page, 'Unexpected undefined page');
+    debug(`pdf to image took ${Date.now() - time} ms`);
+    time = Date.now();
+
+    const ballotBinaryBitmap = imageDataToBinaryBitmap(page, options);
+    debug(`bitmap width: ${ballotBinaryBitmap.width}`);
+    debug(`bitmap height: ${ballotBinaryBitmap.height}`);
+    debug(`image to binary took ${Date.now() - time} ms`);
+    time = Date.now();
+
+    const customChunkedBitmaps = chunkBinaryBitmap(ballotBinaryBitmap);
+    debug(`num chunk rows: ${customChunkedBitmaps.length}`);
+    debug(`binary to chunks took ${Date.now() - time} ms`);
+    time = Date.now();
+
+    await enablePrintPromise;
+    let dotsSkipped = 0;
+    debug(`begin printing ${customChunkedBitmaps.length} chunks`);
+    let i = 0;
+    for (const customChunkedBitmap of customChunkedBitmaps) {
+      debug(`printing chunk ${i}`);
+      if (customChunkedBitmap.empty) {
+        dotsSkipped += 24;
```

Comment:
Magic number.

Reply:
I think this is the line height but I'll confirm this and update

Reply:
24 is the number of dots in the vertical direction when we buffer chunks to print: https://github.com/votingworks/vxsuite/pull/3643/commits/bba9d146154c248c91476d669a65d940c99090e8